packages/ember-glimmer/lib/component-managers/custom.ts

```diff
@@ -1,32 +1,119 @@
-import { ComponentCapabilities, Opaque, Option } from '@glimmer/interfaces';
+import { assert } from '@ember/debug';
+import {
+  ComponentCapabilities,
+  Dict,
+  Opaque,
+  Option,
+  ProgramSymbolTable,
+} from '@glimmer/interfaces';
 import { PathReference, Tag } from '@glimmer/reference';
-import { Arguments, Bounds, CapturedNamedArguments, PrimitiveReference } from '@glimmer/runtime';
+import {
+  Arguments,
+  CapturedArguments,
+  ComponentDefinition,
+  Invocation,
+  PrimitiveReference,
+  WithStaticLayout,
+} from '@glimmer/runtime';
 import { Destroyable } from '@glimmer/util';
-
-import { addChildView } from 'ember-views';
+import { Factory } from 'ember-owner';
+import { OwnedTemplateMeta } from 'ember-views';
 
 import Environment from '../environment';
-import { DynamicScope, Renderer } from '../renderer';
+import { DynamicScope } from '../renderer';
+import RuntimeResolver from '../resolver';
+import { OwnedTemplate } from '../template';
 import { RootReference } from '../utils/references';
 import AbstractComponentManager from './abstract';
-import DefinitionState from './definition-state';
-
-export interface CustomComponentManagerDelegate<T> {
-  version: 'string';
-  create(options: { ComponentClass: T; args: {} }): T;
-  getContext(instance: T): Opaque;
-  update(instance: T, args: {}): void;
-  destroy?(instance: T): void;
-  didCreate?(instance: T): void;
-  didUpdate?(instance: T): void;
-  getView?(instance: T): any;
+
+const CAPABILITIES = {
+  dynamicLayout: false,
+  dynamicTag: false,
+  prepareArgs: false,
+  createArgs: true,
+  attributeHook: false,
+  elementHook: false,
+  createCaller: false,
+  dynamicScope: true,
+  updateHook: true,
+  createInstance: true,
+};
+
+export interface OptionalCapabilities {
+  asyncLifecycleCallbacks?: boolean;
+  destructor?: boolean;
 }
 
-export interface ComponentArguments<T = {}> {
+export function capabilities(
+  managerAPI: '3.4',
+  options: OptionalCapabilities = {}
+): ComponentCapabilityMask {
+  assert('Invalid component manager compatibility specified', managerAPI === '3.4');
+
+  return {
+    asyncLifeCycleCallbacks: !!options.asyncLifecycleCallbacks,
+    destructor: !!options.destructor,
+  };
+}
+
+export interface DefinitionState<ComponentInstance = Opaque> {
+  name: string;
+  ComponentClass: Factory<ComponentInstance>;
+  symbolTable: ProgramSymbolTable;
+  template?: any;
+}
+
+export interface ComponentCapabilityMask {
```

Comment:
Probably just call this `Capabilities`. We probably want to mark the "content" of this private somehow, since from the user's/public API perspective, the function is returning an opaque (and explicitly unstable between versions) thing.